### ISSUE TITLE
Unable to write on osx.

### ISSUE THREAD

**Issue by spambus** [LoganWright/PlistModel#4]
In your documentation it says: "you can call save explicitly using saveInBackgroundWithBlock"
but there is no such function in the code, only "saveInBackgroundWithCompletion", so I guess this must be a typo?
also how do you use this function exactly to save to a plist file?
If I try to use:
[PlistModel plistNamed:@"Info" saveInBackgroundWithCompletion:^(PlistModel *plistModel) {
   plistModel.Test = @"This is a test";    
}];
this doesn't work.
could you perhaps give an example how to use "saveInBackgroundWithCompletion" or how to save a value to a plist file?

**Comment by LoganWright** [comment]
Hi @spambus 

You are correct, that is a typo in the README, sorry about the confusion with that.  All typos aside, you're not really using it correctly.  Save in background assumes that all of the updates have been completed prior to saving.  You'd need to set your `Test` property before you save in the example above.

```ObjC
yourInfoPlist.Test = @"This is a test";
[yourInfoPlist saveInBackgroundWithCompletion:^(PlistModel *plistModel) {

}];
```

However, there's a secondary problem.  I noticed that you're using `@"Info"` and loading the default Plist.  Are you loading the bundled Plist? If so, you won't be able to add new values.  This is a limitation of iOS, and mentioned explicitly in the README.  Make sure you never try to edit an immutable / bundled Plist.  

I suggest looking at the examples bundled with the project to get a bit more example on how to use it.

Thanks for checking it out and let me know if you have any other problems :)

Best,

Logan

**Comment by spambus** [comment]
Hi

actually I'm using your code on OSX, not on iOS

reading from a plist works just fine

I did set:
@property (strong, nonatomic) NSString * Test;
I also tried using another plist besides Info.plist
but that also doesn't work.

and:

[yourInfoPlist saveInBackgroundWithCompletion:^(PlistModel *plistModel) {

}];

also doesn't work :S





2015-03-01 15:07 GMT+01:00 Logan Wright <notifications@github.com>:

**Comment by LoganWright** [comment]
Did you try another Plist that was included in your bundle?  If the Plist is in your Xcode project, you won't be able to edit it.  The only way that a plist can be modified is to create it dynamically in a directory where you have readwrite access.  

Are you definitely creating it dynamically, or are you reading from a Plist in your project and trying to edit that one?

**Comment by spambus** [comment]
I have write acces to the plist

I compile the code and then place the app in the /Applications folder,
where the plist file has read/write acces, but also in the folder where it
is compiled the plist has read/write acces

I don't think there is a write restriction on OSX to Info.plist btw... but
also tested using Data.plist

But the problem is that the saving code gives a compile error, so i can't
test it if it actually works.

I only started learning objective c 2 days ago, I'm more familiar with
c++,php and python

It probably comes down to not having the exact needed code to save to a
plist

for example I use this to read the boolean value from a plist to enable
sound in my program and set a checkbox in the program accordingly when the
program starts:

[PlistModel plistNamed:@"Info" inBackgroundWithBlock:^(PlistModel
*plistModel) {

       sound = plistModel.Sound;

      if (sound == YES){

            [_SOnOff setState:1];

       }

       else{

            [_SOnOf setState: 0];

        }

 }];
Now I just need to figure out how to also write to a plist file

on the checkbutton code I have:

- (IBAction)Sound:(id)sender {

    if (sound == YES){

        NSBeep();

    }

    if ([(NSButton*)sender state] == NSOnState){

        sound = YES;

    }

    else{

        sound = NO;

    }

}

but it needs the exact code to save to a plist file

if I add:

    [PlistModel plistNamed:@"Info" inBackgroundWithBlock:^(PlistModel
*plistModel) {

        plistModel.Sound = sound

    }];

thats not it

also

plistModel.Sound = sound;
[yourInfoPlist saveInBackgroundWithCompletion:^(PlistModel *plistModel) {

}];

is not working, because it is not the exact needed code to save

I just need an exact example in order to understand it

thanx...


2015-03-01 15:30 GMT+01:00 Logan Wright <notifications@github.com>:

**Comment by LoganWright** [comment]
No problem,

Is that the exact code you're calling, because after looking closer, that is not how the saveInBackground method looks.  Have you tried this:

```ObjC
plistModel.Sound = sound;
[plistModel saveInBackgroundWithCompletion:^(BOOL successful) {

}];
```

**Comment by spambus** [comment]
I'm sorry to bother you again but I can't seem to get it working
with:

[plistModel saveInBackgroundWithCompletion:^(BOOL successful) {

}];

the code compiles without errors, so I was finally able to test it.

apparently on OSX you also cannot use Info.plist to write to

so I used Data.plist

If Data.plist is inside the compiled app bundle this also doesn't
work, then it can only read values from it and console log gives an
error:

Bundled Plists are immutable on a RealDevice, New values will not save!

So I put the Data.plist file in my Documents folder

then it doesn't give a error in the console log

and reads the values from it, but still doesn't save values, alltough
there is write acces to the file

btw the above code part needs to be inside the other code like this to
compile without errors:

[PlistModel plistNamed:@"Data" inBackgroundWithBlock:^(PlistModel *plistModel) {

            plistModel.Sound = sound;

            NSLog(@"Geluid = %@", plistModel.Sound ? @"YES" : @"NO");

            [plistModel saveInBackgroundWithCompletion:^(BOOL
successful) {

            }];

        }];

}

but it doesn't save the values, I also tried:

[PlistModel plistNamed:@"Data" inBackgroundWithBlock:^(PlistModel *plistModel) {

            plistModel.Sound = sound; // Here first

            NSLog(@"Geluid = %@", plistModel.Sound ? @"YES" : @"NO");

            [plistModel saveInBackgroundWithCompletion:^(BOOL successful) {

                                  plistModel.Sound = sound;  // and
also here to test if it needs to be inside here

            }];

        }];

}

but that also doesn't work.


2015-03-01 16:51 GMT+01:00 Logan Wright <notifications@github.com>:

**Comment by LoganWright** [comment]
Can you show me where `Sound` is declared?

**Comment by spambus** [comment]
in PlistModel.h
@property BOOL Sound;

**Comment by LoganWright** [comment]
I'm not sure what's going on at this point, would it be possible for you to add me as a collaborator on your project so I can look at it?

**Comment by spambus** [comment]
Done

**Comment by LoganWright** [comment]
@spambus - The issue appears to be a problem w/ osx. This line here:

```ObjC
BOOL successful = NO;
successful = [dictionary writeToFile:path atomically:YES];
```

Is always returning no.  From what I've gathered, it sounds like there are more read write permissions that I'm not aware of in an osx app. 

I don't have much experience in osx, and I can't give you a good timeline of when I'll be able to dig in more than I have already.  I will try to find someone with more experience w/ osx who can help me figure it out.

If you have more experience with osx, perhaps you'd be able to fix the area I mentioned.  

Sorry I don't have better news, if I find time to look at this more, I'll update you.  In meantime, perhaps something like Mantle or Realm might help you.

Best,

Logan

**Comment by LoganWright** [comment]
@spambus 

I was able to resolve the issue somewhat, I'm not sure why, but it's allowing me to write when I archive as data.  This should provide the desired effect for the time being.  I also submitted a PR to your project with the updates included in it.  I recommend subclassing to your own model so it's easier to update w/o messing w/ the source code.

Let me know if you have any more issues.

Best,

Logan
